[otbn] Fix some issues in modexp assembly

imphil opened the pull request:
The modexp assembly doesn't currently run on the new simulator. Some
problems are simulator-related, others are bugs in the code. I'm still
going through all of them, but here are two fixes which are clearly
wrong in the assembly.

* LW expects offsets in bytes, not in 32b words.
* In contrast to dcrypto1, routines need to end with an explicit ECALL
  instruction.

imphil commented:
@felixmiller This also shows a simulator mismatch. The old simulator assumes the behavior of the original file.

felixmiller commented:
@imphil you should be right about the offsets. Testing with your changes and the "old" sim now.

re `ecall`: The `mulx` and `mul1` routines also get externally called. `mulx` gets both internally and externally called, so I guess we need a wrapper for this. `mul1` should be good with an `ecall`

felixmiller commented:
@imphil: I think you missed t[hese here](https://github.com/lowRISC/opentitan/blob/40031c8cf9b019dfe7b72af1b03bd09031f838e5/sw/otbn/code-snippets/modexp.S#L442)

felixmiller commented on sw/otbn/code-snippets/modexp.S:
```diff
@@ -336,19 +336,19 @@ JAL x1, mul1_exp
-LW x21, 13(x0)
-LW x22, 14(x0)
-LW x23, 15(x0)
+LW x16, 24(x0)
```
```suggestion
LW x16, 32(x0)
```

imphil replied:
Fixed

imphil commented:
I fixed the LW instructions I missed, and added an ecall to the mul1 function. I left mulx as-is for now, since I don't have a way to test the external caller. I'd leave that for a follow-up.

felixmiller reviewed (changes requested):
Fixed a bug in the old simulator in the `lw` instruction.
Works now with your changes with the small modification in line 339 and 442++.

felixmiller reviewed (approved):
LGTM and runs again in the fixed old simulator

rswarbrick reviewed (approved):
Looks good to me.

imphil commented:
Rebased to solve the merge conflict. Merging since this code isn't used in CI anyways (yet).

imphil commented:
Rebased to solve the merge conflict. Merging since this code isn't used in CI anyways (yet) and lint steps passed.
